src/protosocket/cache/address_provider.rs

```diff
@@ -64,85 +88,242 @@ impl std::fmt::Display for RefreshError {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum AddressProvider {
+    Refreshing {
+        addresses: Arc<RwLock<AddressState>>,
+        _background_loader: BackgroundAddressLoader,
+    },
+    Static {
+        address: SocketAddr,
+    },
+}
+
 impl AddressProvider {
-    /// Looks for an address list from the provided endpoint.
-    #[allow(clippy::expect_used)]
-    pub fn new(credential_provider: CredentialProvider) -> Self {
+    pub async fn new(
+        credential_provider: CredentialProvider,
+        runtime: tokio::runtime::Handle,
+    ) -> MomentoResult<Self> {
+        if credential_provider.use_endpoints_http_api {
+            Self::new_with_refresh(credential_provider, runtime).await
+        } else {
+            Self::new_with_static_address(credential_provider)
+        }
+    }
+
+    fn new_with_static_address(credential_provider: CredentialProvider) -> MomentoResult<Self> {
+        let address_string =
+            if matches!(credential_provider.endpoint_security, EndpointSecurity::Tls) {
+                // TLS: make sure https:// is stripped and force port 9004
+                let endpoint = credential_provider
+                    .cache_endpoint
+                    .strip_prefix("https://")
+                    .unwrap_or(&credential_provider.cache_endpoint);
+                let hostname = endpoint.split(':').next().unwrap_or(endpoint);
```

Comment:
this has the behavior of stripping a url path when a port is present, but not stripping that path when there is no port in the url. Is that intentional?